source/common/router/config_impl.cc

```diff
@@ -698,48 +692,66 @@ void RouteEntryImplBase::finalizeRequestHeaders(Http::RequestHeaderMap& headers,
-  } else {
-    // Most specific mutations (route-level) take precedence by being applied
-    // last: if a header is specified at all levels, the last one applied wins.
-    mergeTransforms(transforms,
-                    vhost_.globalRouteConfig().responseHeaderParser().getHeaderTransforms(
-                        stream_info, do_formatting));
-    mergeTransforms(transforms,
-                    vhost_.responseHeaderParser().getHeaderTransforms(stream_info, do_formatting));
-    mergeTransforms(transforms,
-                    response_headers_parser_->getHeaderTransforms(stream_info, do_formatting));
+  for (const HeaderParser* header_parser : getHeaderParsers(
+           /*get_request_header_parser=*/false,
+           /*specificity_ascend=*/vhost_.globalRouteConfig().mostSpecificHeaderMutationsWins())) {
+    // Later evaluated header parser wins.
+    mergeTransforms(transforms, header_parser->getHeaderTransforms(stream_info, do_formatting));
+  }
+  return transforms;
+}
+
+Http::HeaderTransforms
+RouteEntryImplBase::requestHeaderTransforms(const StreamInfo::StreamInfo& stream_info,
+                                            bool do_formatting) const {
+  Http::HeaderTransforms transforms;
+  for (const HeaderParser* header_parser : getHeaderParsers(
+           /*get_request_header_parser=*/true,
+           /*specificity_ascend=*/vhost_.globalRouteConfig().mostSpecificHeaderMutationsWins())) {
+    // Later evaluated header parser wins.
+    mergeTransforms(transforms, header_parser->getHeaderTransforms(stream_info, do_formatting));
   }
   return transforms;
 }
 
+std::vector<const HeaderParser*>
+RouteEntryImplBase::getHeaderParsers(bool get_request_header_parser,
+                                     bool specificity_ascend) const {
+  std::vector<const HeaderParser*> parsers;
+  const HeaderParser* global_route_config_header_parser =
+      get_request_header_parser ? &vhost_.globalRouteConfig().requestHeaderParser()
+                                : &vhost_.globalRouteConfig().responseHeaderParser();
+  const HeaderParser* vhost_header_parser =
+      get_request_header_parser ? &vhost_.requestHeaderParser() : &vhost_.responseHeaderParser();
+  const HeaderParser* route_header_parser =
+      get_request_header_parser ? request_headers_parser_.get() : response_headers_parser_.get();
+  if (specificity_ascend) {
+    // Appends user-specified request headers from least to most specific: global connection manager
+    // level headers, virtual host level headers and finally route-level headers.
+    parsers.push_back(global_route_config_header_parser);
+    parsers.push_back(vhost_header_parser);
+    parsers.push_back(route_header_parser);
+  } else {
+    // Appends user-specified request headers from most to least specific.
+    parsers.push_back(route_header_parser);
+    parsers.push_back(vhost_header_parser);
+    parsers.push_back(global_route_config_header_parser);
+  }
+  return parsers;
+}
```

Comment:
``
```suggestion
template<bool FOR_REQUEST_HEADERS> absl::InlinedVector<const HeaderParser*, 3>
RouteEntryImplBase::getHeaderParsers(bool specificity_ascend) const {
  const HeaderParser* global_route_config_header_parser = nullptr;
  const HeaderParser* vhost_header_parser = nullptr;
  const HeaderParser* route_header_parser = nullptr;
  
  if constexpr (FOR_REQUEST_HEADERS) {
    global_route_config_header_parser = &vhost_.globalRouteConfig().requestHeaderParser();
    vhost_header_parser =  &vhost_.requestHeaderParser();
    route_header_parser = request_headers_parser_.get()
  } else {
    global_route_config_header_parser = &vhost_.globalRouteConfig().responseHeaderParser();
    vhost_header_parser =  &vhost_.responseHeaderParser();
    route_header_parser = response_headers_parser_.get()
  }
  return specificity_ascend ? {global_route_config_header_parser,  vhost_header_parser, route_header_parser} : 
                              {route_header_parser,  vhost_header_parser, global_route_config_header_parser};
  }
}
```

Reply:
Hmm. Thanks for the example! That makes sense. Didn't know `Non-type template parameters` is a thing. So I assume the improvement here is not about writing less code but more about efficiency (because the bool now are allocated statically at compiling time so it is faster at run time)? 

I have split this function into getRequestHeaderParsers and getResponseHeaderParsers as per the other comment, so the `bool get_request_header_parser` is gone.  Do you think I should make template based on `bool specificity_ascend`?

Reply:
@yanavlasov Thanks. Done.

Reply:
There is no reason in using templates with bool parameter here as it actually does not reduce the amount of code you need write. But it does reduce readability.

Reply:
> I think it could be simpler and more readable to add two functions `getRequestHeaderParsers` and `getResponseHeaderParsers` and one extra helper that deals with the `specificity_ascend`.

This is also great.